### PR TITLE
Optimistic lock for users

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

Optimistic locking on user resource.
On the http side it's done via http conditional requests, using etag and if-match header.
Openapi updated with the proper headers for requests and responses.
Recording 2024-05-06 at 16 32 07

Swagger playground updated.

Frontend updated.

## How was this tested?

Automated tests

### CODE REVIEW

File: lib/trento/users/user.ex [hunk 1]
```diff
@@ -51,6 +52,8 @@ defmodule Trento.Users.User do
     |> validate_password()
     |> custom_fields_changeset(attrs)
     |> lock_changeset(attrs)
+    |> cast(attrs, [:lock_version])
+    |> optimistic_lock(:lock_version)
```
Comment (arbulu89) [inline lib/trento/users/user.ex:56]: :heart: ecto!

File: test/trento/users_test.exs [hunk 2]
```diff
@@ -250,6 +250,22 @@ defmodule Trento.UsersTest do
                Users.update_user(%User{id: 1}, %{fullname: "new fullname"})
     end
 
+    test "update_user/2 return stale error when the lock version in the update is not valid" do
```
Comment (arbulu89) [inline test/trento/users_test.exs:253]: `returns` with `s`

**Review by arbulu89 — APPROVED** [review]
Perfection!

**Review by EMaksy — APPROVED** [review]
Hey man, great, left just some last questions :+1:

File: lib/trento_web/openapi/v1/schema/precondition_failed.ex [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  OpenApiSpex.schema(%{
+    title: "PreconditionFailed",
+    type: :object,
+    additionalProperties: false,
+    properties: %{
+      errors: %Schema{
+        type: :array,
+        items: %Schema{
+          type: :object,
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento_web/openapi/v1/schema/precondition_failed.ex:18]: You spotted a tech debt for all our error schemas!

We don't have in any of them, basically it's just a copy paste

File: assets/js/lib/api/users.js [hunk 4]
```diff
@@ -6,6 +6,11 @@ export const getUser = (userID) => get(`/users/${userID}`);
 
 export const createUser = (payload) => post('/users', payload);
 
-export const editUser = (userID, payload) => patch(`/users/${userID}`, payload);
+export const editUser = (userID, payload, version) =>
+  patch(`/users/${userID}`, payload, {
+    headers: {
+      'if-match': version,
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply assets/js/lib/api/users.js:12]: It's an http standard header, like the etag

File: test/trento/users_test.exs (hunk 2 above)
Comment (CDimonaco) [reply test/trento/users_test.exs:253]: done!

**Pull request merged**